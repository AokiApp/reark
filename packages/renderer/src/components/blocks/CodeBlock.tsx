import { BlockInnerComponent } from "../../types";
import { CODE_LANGUAGE } from "../../constants/codeLanguage";
import { Text } from "./Text";

export const CodeBlock: BlockInnerComponent = ({ block }) => {
  const code = block.code;
  if (!code) {
    return null;
  }
  const codeLanguage = code?.style?.language
    ? CODE_LANGUAGE[code.style.language]
    : "Unknown";

  return (
    <div className="reark-codeblock">
      <div className="reark-codeblock__language">{codeLanguage}</div>
      <pre className="reark-codeblock__pre">
        <Text elements={code.elements} style={code.style} />
      </pre>
    </div>
  );
};
// todo : copy button
