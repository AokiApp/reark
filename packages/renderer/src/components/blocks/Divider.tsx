import { BlockInnerComponent } from "../../types";

export const Divider: BlockInnerComponent = () => {
  return <hr className="reark-divider" />;
};
