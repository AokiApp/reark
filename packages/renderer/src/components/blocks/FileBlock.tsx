import { BlockInnerComponent } from "../../types";

/**
 * FileBlock: Renders a file block (block_type: 23).
 * - Shows file name and download link if available.
 * - This block is usually paired with a ViewBlock (block_type: 33) for visual representation.
 * - If rendered directly, provides basic file info.
 */
export const FileBlock: BlockInnerComponent = ({ block }) => {
  const file = block.file;
  if (!file) {
    return (
      <div className="reark-file-block reark-file-block--empty">
        No file data
      </div>
    );
  }

  return (
    <div className="reark-file-block">
      <div>
        <strong>File:</strong> {file.name || "(no name)"}
      </div>
      {file.token ? (
        <a
          href={`https://open.feishu.cn/drive/file/${file.token}`}
          target="_blank"
          rel="noopener noreferrer"
          className="reark-file-block__file-link"
        >
          Download / Open
        </a>
      ) : (
        <span className="reark-file-block__file-link--disabled">
          (No file token)
        </span>
      )}
    </div>
  );
};
