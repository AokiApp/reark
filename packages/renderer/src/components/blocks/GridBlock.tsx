import { BlockInnerComponent } from "../../types";
import { BlockComponent } from "../BlockComponent";

/**
 * GridBlock: Renders a grid container with a specified number of columns.
 * Each child is expected to be a GridColumnBlock.
 */
export const GridBlock: BlockInnerComponent = ({ block }) => {
  const grid = block.grid;
  if (!grid || !block.children || !Array.isArray(block.children)) {
    return null;
  }

  // column_size: number of columns (2-4)
  //const columnSize = grid.column_size || 2;

  // Render children by block reference (blockId), allowing for block reference resolution.
  return (
    <div className="reark-grid-wrapper">
      <div
        className="reark-grid"
        style={{
          display: "flex",
          //gridTemplateColumns: `repeat(${columnSize}, 1fr)`,
          gap: "16px",
        }}
      >
        {block.children.map((childId) => (
          // BlockComponent resolves and renders the block by its ID (block reference)
          <BlockComponent key={childId} blockId={childId} />
        ))}
      </div>
    </div>
  );
};
