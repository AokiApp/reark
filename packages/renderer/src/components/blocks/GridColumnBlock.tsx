import { BlockInnerComponent } from "../../types";
import { BlockComponent } from "../BlockComponent";

/**
 * GridColumnBlock: Renders a single column in a grid, with optional width ratio.
 * Children can be any block type.
 */
export const GridColumnBlock: BlockInnerComponent = ({ block }) => {
  const gridColumn = block.grid_column;
  if (!block.children || !Array.isArray(block.children)) {
    return null;
  }

  // width_ratio: ratio of this column to the whole grid (1-99)
  const widthRatio = gridColumn?.width_ratio || 1;

  // Render children by block reference (blockId), allowing for block reference resolution.
  return (
    <div
      className="reark-grid-column"
      style={{
        flex: widthRatio,
        minWidth: 0,
        display: "flex",
        flexDirection: "column",
        gap: "8px",
      }}
    >
      {block.children.map((childId) => (
        // BlockComponent resolves and renders the block by its ID (block reference)
        <BlockComponent key={childId} blockId={childId} />
      ))}
    </div>
  );
};
