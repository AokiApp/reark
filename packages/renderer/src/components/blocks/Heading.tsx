import { BlockInnerComponent } from "../../types";
import { Text } from "./Text";

type HeadingLevel = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

const HeadingBase: BlockInnerComponent<{
  level: HeadingLevel;
}> = ({ block, level }) => {
  const headingKey = `heading${level}` as const;
  const headingData = block[headingKey];

  if (!headingData?.elements) {
    return null;
  }

  return (
    <div
      className={`reark-heading reark-heading--${level}`}
      id={block.block_id}
    >
      <Text elements={headingData.elements} style={headingData.style} />
    </div>
  );
};

export const Heading1: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={1} />
);
export const Heading2: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={2} />
);
export const Heading3: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={3} />
);
export const Heading4: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={4} />
);
export const Heading5: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={5} />
);
export const Heading6: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={6} />
);
export const Heading7: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={7} />
);
export const Heading8: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={8} />
);
export const Heading9: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={9} />
);
