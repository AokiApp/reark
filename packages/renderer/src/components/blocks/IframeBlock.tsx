import React from "react";
import { BlockInnerComponent } from "../../types";

function decodeUrl(url: string): string {
  try {
    return decodeURIComponent(url);
  } catch {
    return url;
  }
}

export const IframeBlock: BlockInnerComponent = ({ block }) => {
  const component = block.iframe?.component;
  if (!component) return null;

  const { iframe_type, url } = component;
  const decodedUrl = decodeUrl(url);

  if (iframe_type === 15) {
    // YouTube埋め込み
    // URL例: https://www.youtube.com/watch?v=xxxx
    // 埋め込み用URLに変換
    let embedUrl = "";
    const match = decodedUrl.match(
      /(?:youtube\.com\/watch\?v=|youtu\.be\/)([A-Za-z0-9_-]+)/,
    );
    if (match && match[1]) {
      embedUrl = `https://www.youtube.com/embed/${match[1]}`;
    } else {
      embedUrl = decodedUrl;
    }
    return (
      <div className="reark-iframe">
        <iframe
          src={embedUrl}
          allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
          allowFullScreen
          title="YouTube Video"
        />
      </div>
    );
  } else {
    // 汎用iframe
    return (
      <div className="reark-iframe">
        <iframe
          src={decodedUrl}
          allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
          allowFullScreen
          title="Embedded Content"
        />
      </div>
    );
  }
};
