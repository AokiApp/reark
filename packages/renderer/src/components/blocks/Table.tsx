import { BlockInnerComponent } from "../../types";
import { BlockComponent } from "../BlockComponent";

/**
 * Table renderer for Lark TableBlock.
 * - Supports header row and column.
 * - Supports cell merging (colSpan, rowSpan) via merge_info.
 * - Renders each cell using BlockComponent.
 */
export const Table: BlockInnerComponent = ({ block }) => {
  // Defensive: If no table data, render nothing
  const table = block.table;
  if (!table || !table.property) return null;

  const { row_size, column_size, header_row, header_column } = table.property;
  const cells = table.cells || [];

  // Build 2D array of cell IDs: [row][col]
  const cellGrid: (string | null)[][] = [];
  for (let r = 0; r < row_size; r++) {
    const row: (string | null)[] = [];
    for (let c = 0; c < column_size; c++) {
      row.push(cells[r * column_size + c] || null);
    }
    cellGrid.push(row);
  }

  // Build a 2D array to track merged cells and their spans
  // merge_info is a flat array, same order as cells
  // For each cell, if merge_info exists, set colSpan/rowSpan; otherwise, default to 1
  type MergeCell = {
    colSpan: number;
    rowSpan: number;
    covered: boolean; // true if this cell is covered by a previous cell's span
  };
  const mergeGrid: MergeCell[][] = [];
  for (let r = 0; r < row_size; r++) {
    const row: MergeCell[] = [];
    for (let c = 0; c < column_size; c++) {
      const idx = r * column_size + c;
      const info = table.property.merge_info?.[idx];
      row.push({
        colSpan: info?.col_span ?? 1,
        rowSpan: info?.row_span ?? 1,
        covered: false,
      });
    }
    mergeGrid.push(row);
  }

  // Mark covered cells (those that should be skipped due to merge)
  for (let r = 0; r < row_size; r++) {
    for (let c = 0; c < column_size; c++) {
      const { colSpan, rowSpan } = mergeGrid[r][c];
      if (colSpan > 1 || rowSpan > 1) {
        for (let dr = 0; dr < rowSpan; dr++) {
          for (let dc = 0; dc < colSpan; dc++) {
            if (dr === 0 && dc === 0) continue; // skip the top-left cell
            const rr = r + dr;
            const cc = c + dc;
            if (rr < row_size && cc < column_size) {
              mergeGrid[rr][cc].covered = true;
            }
          }
        }
      }
    }
  }

  // Render table
  return (
    <div className="reark-table-wrapper">
      <table className="reark-table">
        <tbody>
          {cellGrid.map((row, r) => (
            <tr key={r}>
              {row.map((cellId, c) => {
                const merge = mergeGrid[r][c];
                if (merge.covered || !cellId) return null;

                // Determine if this cell is a header cell
                const isHeader =
                  (header_row && r === 0) || (header_column && c === 0);

                const CellTag = isHeader ? "th" : "td";

                return (
                  <CellTag
                    key={c}
                    colSpan={merge.colSpan > 1 ? merge.colSpan : undefined}
                    rowSpan={merge.rowSpan > 1 ? merge.rowSpan : undefined}
                    className={[
                      isHeader ? "reark-table__th" : "reark-table__td",
                      merge.colSpan > 1 || merge.rowSpan > 1
                        ? "reark-table__merged"
                        : "",
                    ].join(" ")}
                  >
                    <BlockComponent blockId={cellId} />
                  </CellTag>
                );
              })}
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};
