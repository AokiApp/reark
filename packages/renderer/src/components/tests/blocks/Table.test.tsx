import { render, screen } from "@testing-library/react";
import { Table } from "../../blocks/Table";
import { BlockStoreProvider } from "../../../contexts/BlockStoreContext";

// ダミーのTableブロック（最小限）

describe("Table block", () => {
  it("renders table and unsupported cells", () => {
    const tableBlock = {
      block_id: "dummy-table",
      block_type: 21,
      table: {
        cells: ["cell-1", "cell-2", "cell-3", "cell-4"],
        property: {
          column_size: 2,
          row_size: 2,
          header_row: true,
          header_column: false,
          column_width: [100, 100],
          merge_info: [
            { col_span: 1, row_span: 1 },
            { col_span: 1, row_span: 1 },
            { col_span: 1, row_span: 1 },
            { col_span: 1, row_span: 1 },
          ],
        },
      },
    };

    const cellBlock1 = { block_id: "cell-1", block_type: 97 };
    const cellBlock2 = { block_id: "cell-2", block_type: 97 };
    const cellBlock3 = { block_id: "cell-3", block_type: 97 };
    const cellBlock4 = { block_id: "cell-4", block_type: 97 };

    const { container } = render(
      <BlockStoreProvider
        items={[tableBlock, cellBlock1, cellBlock2, cellBlock3, cellBlock4]}
      >
        <Table block={tableBlock} />
      </BlockStoreProvider>,
    );
    // .reark-table-wrapper要素が存在すること
    const wrapper = container.querySelector(".reark-table-wrapper");
    expect(wrapper).not.toBeNull();
    // Unsupported block type: 97 の表示が4つあること
    const unsupported = screen.getAllByText(/Unsupported block type: 97/);
    expect(unsupported.length).toBe(4);
    // スナップショットテスト
    expect(container).toMatchSnapshot();
  });
});
