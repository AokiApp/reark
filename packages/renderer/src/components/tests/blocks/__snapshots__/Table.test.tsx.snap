// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table block renders table and unsupported cells 1`] = `
<div>
  <div
    class="reark-table-wrapper"
  >
    <table
      class="reark-table"
    >
      <tbody>
        <tr>
          <th
            class="reark-table__th "
          >
            <div
              class="reark-unsupported-block"
            >
              <strong>
                Unsupported block type: 
                97
              </strong>
            </div>
          </th>
          <th
            class="reark-table__th "
          >
            <div
              class="reark-unsupported-block"
            >
              <strong>
                Unsupported block type: 
                97
              </strong>
            </div>
          </th>
        </tr>
        <tr>
          <td
            class="reark-table__td "
          >
            <div
              class="reark-unsupported-block"
            >
              <strong>
                Unsupported block type: 
                97
              </strong>
            </div>
          </td>
          <td
            class="reark-table__td "
          >
            <div
              class="reark-unsupported-block"
            >
              <strong>
                Unsupported block type: 
                97
              </strong>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`renders table with no cells 1`] = `
<div>
  <div
    class="reark-table-wrapper"
  >
    <table
      class="reark-table"
    >
      <tbody>
        <tr />
        <tr />
      </tbody>
    </table>
  </div>
</div>
`;
